File: packages/sdk-py/theatrical/types/subscription.py
"""Subscription types."""

from __future__ import annotations

from enum import Enum
from typing import Optional

from pydantic import BaseModel


class SubscriptionInterval(str, Enum):
    MONTHLY = "monthly"
    ANNUAL = "annual"


class SubscriptionStatus(str, Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    CANCELLED = "cancelled"
    EXPIRED = "expired"
    PENDING = "pending"


class BenefitCategory(str, Enum):
    BOOKING = "booking"
    CONCESSION = "concession"
    COMPANION = "companion"
    PRIORITY = "priority"
    EXCLUSIVE = "exclusive"


class SubscriptionBenefit(BaseModel):
    id: str
    category: BenefitCategory
    name: str
    description: str
    uses_per_period: Optional[int] = None
    active: bool


class SubscriptionPlan(BaseModel):
    id: str
    name: str
    description: str
    price: float
    currency: str
    interval: SubscriptionInterval
    bookings_included: Optional[int] = None
    benefits: list[SubscriptionBenefit]
    available: bool
    minimum_term_months: Optional[int] = None


class SubscriptionUsage(BaseModel):
    subscription_id: str
    member_id: str
    period_start: str
    period_end: str
    bookings_used: int
    bookings_included: Optional[int] = None
    bookings_remaining: Optional[int] = None
    benefit_usage: dict[str, int] = {}


class MemberSubscription(BaseModel):
    id: str
    plan_id: str
    plan: Optional[SubscriptionPlan] = None
    member_id: str
    status: SubscriptionStatus
    start_date: str
    renewal_date: Optional[str] = None
    cancelled_at: Optional[str] = None
    expires_at: Optional[str] = None
    auto_renew: bool
    usage: Optional[SubscriptionUsage] = None


class SuspendSubscriptionInput(BaseModel):
    resume_date: Optional[str] = None
    reason: Optional[str] = None


class CancelSubscriptionInput(BaseModel):
    immediate: Optional[bool] = None
    reason: Optional[str] = None

File: packages/sdk-csharp/src/Theatrical.Sdk/Types/Subscription.cs
namespace Theatrical.Sdk.Types;

public sealed record SubscriptionPlan
{
    public required string Id { get; init; }
    public required string Name { get; init; }
    public required decimal PricePerMonth { get; init; }
    public string? Description { get; init; }
    public string[]? Benefits { get; init; }
    public bool IsAvailable { get; init; } = true;
}

public sealed record MemberSubscription
{
    public required string MemberId { get; init; }
    public required string PlanId { get; init; }
    public required string Status { get; init; }
    public string? StartDate { get; init; }
    public string? EndDate { get; init; }
    public string? NextBillingDate { get; init; }
}

public sealed record SubscriptionUsage
{
    public required string MemberId { get; init; }
    public required string PlanId { get; init; }
    public int BookingsUsed { get; init; }
    public int? BookingsRemaining { get; init; }
    public string? PeriodStart { get; init; }
    public string? PeriodEnd { get; init; }
    public BenefitUsage[]? Benefits { get; init; }
}

public sealed record BenefitUsage
{
    public required string BenefitId { get; init; }
    public required string Name { get; init; }
    public int Used { get; init; }
    public int? Limit { get; init; }
}

public sealed record BenefitEligibility
{
    public bool Eligible { get; init; }
    public int? UsesRemaining { get; init; }
    public string? Reason { get; init; }
}

public sealed record SuspendSubscriptionInput
{
    public string? ResumeDate { get; init; }
    public string? Reason { get; init; }
}

public sealed record CancelSubscriptionInput
{
    public bool Immediate { get; init; }
    public string? Reason { get; init; }
}
